### ISSUE TITLE
cue/load: do not include module ancestor files in the same package by default; make optional

### ISSUE THREAD

**Issue by myitcv** [cue-lang/cue#2000]
Currently, per https://cuelang.org/docs/concepts/packages/:

> Within a module, all `.cue` files with the same package name are part of the same package. A package is evaluated within the context of a certain directory. Within this context, only the files belonging to that package in that directory and its ancestor directories within the module are combined. We call that an instance of a package.

As a demonstration of the current behaviour, the following test passes:

```
exec cue eval ./...
cmp stdout stdout.golden

-- cue.mod/module.cue --
module: "example.com"
-- root.cue --
package example

x: a: 5
-- subdir/subdir.cue --
package example

x: b: 4
-- stdout.golden --
x: {
    a: 5
}
// ---
x: {
    a: 5
    b: 4
}
```

Feedback from many users has shown this to be a confusing behaviour by default in certain situations: 

* I was expecting a Go-like package experience, this is not that.
* I have multiple services defined within a monorepo, each defining its own configuration as part of a `config` package local to that service. This pattern does not work in that context.
* ... (please reply below with further examples of how/where this is frustrating)

This umbrella issue tracks:

1. Changing the default behaviour of CUE to _not_ include ancestor files in the same package. From experience, this is the most commonly expected default.
2. Proving a configuration option to enable the current default behaviour.

Concretely, the new proposed default would result in the following test passing:

```
exec cue eval ./...
cmp stdout stdout.golden

-- cue.mod/module.cue --
module: "example.com"
-- root.cue --
package example

x: a: 5
-- subdir/subdir.cue --
package example

x: b: 4
-- stdout.golden --
x: {
    a: 5
}
// ---
x: {
    b: 4
}
```

### Open questions

* Where the configuration option would be set. Likely somewhere within `cue.mod/module.cue`. 
* Whether this would be an all-or-nothing configuration option, or whether there would be a package-level granularity. 
* The use of a CUE directive in a `cue.mod/module.cue` file (or package) to enable this behaviour from a specific release onwards (separate issue to follow).

**Comment by myitcv** [comment]
As part of this issue we should also consider how/where `cue.mod/{usr,gen}` fit into the picture.

**Comment by myitcv** [comment]
Linking to https://github.com/cue-lang/cue/issues/2004, because a change in the default should be gated on a language/toolchain version.

**Comment by mvdan** [comment]
Linking to https://github.com/cue-lang/cue/issues/1781 and https://github.com/cue-lang/cue/issues/3973 as they are two other issues with examples of users getting very confused by package instances, understandably.

@cuematthew and I also realised that this is confusing in the context of `cue fmt`; it loads "packages" but, unlike almost every other command, it does not load "package instances". For example, `cue export .` will load ancestor packages with the same name, but `cue fmt` will not; it will only ever format CUE files in the current directory.

So, if/when we change the default and make "package instance" loading configurable, we should ensure that the docs are consistent in explaining which commands actually use this mechanism.

**Comment by mvdan** [comment]
Also see https://github.com/cue-lang/cue/issues/2486; I would argue that we should close this issue in favor of that newer one, as it contains a bit more research into the use cases we've encountered, and it has also had more attention from users in the form of reactions. @myitcv thoughts?

**Comment by mvdan** [comment]
Closing in favor of #2486; I keep finding this issue when I do a quick open issue search, and missing the newer and more active thread until I see my last comment. The newer issue will be easier to discover if it's the only one that remains open.
